### PR TITLE
Refactor syntax/semantic parsing interfaces

### PR DESCRIPTION
**Pull request by wixette**

### CODE REVIEW

File: csharp/src/SeedLang/X/VisitorHelper.cs [hunk 1]
```diff
@@ -310,12 +311,12 @@ internal class VisitorHelper {
             exprs[i + additionalParam] = expr;
           }
           if (i < commaNodes.Length) {
-            AddSyntaxToken(SyntaxType.Symbol,
+            AddSemanticToken(TokenType.Symbol,
                            CodeReferenceUtils.RangeOfToken(commaNodes[i].Symbol));
```
Comment (codingpotato) [inline csharp/src/SeedLang/X/VisitorHelper.cs:315]:
```suggestion
                             CodeReferenceUtils.RangeOfToken(commaNodes[i].Symbol));
```

File: csharp/src/SeedLang/X/SeedPython.cs [hunk 2]
```diff
@@ -22,34 +22,72 @@
...
+      { SeedPythonParser.ADD_ASSIGN, TokenType.Operator },
+      { SeedPythonParser.SUBSTRACT_ASSIGN, TokenType.Operator },
+      { SeedPythonParser.MULTIPLY_ASSIGN, TokenType.Operator },
+      { SeedPythonParser.DIVIDE_ASSIGN, TokenType.Operator },
+      { SeedPythonParser.MODULO_ASSIGN, TokenType.Operator },
+      { SeedPythonParser.OPEN_PAREN, TokenType.OpenParenthesis },
+      { SeedPythonParser.CLOSE_PAREN, TokenType.CloseParenthesis },
+      { SeedPythonParser.OPEN_BRACK, TokenType.OpenBracket },
+      { SeedPythonParser.CLOSE_BRACK, TokenType.CloseBracket },
+      // Ignore: OPEN_BRACE
```
Comment (codingpotato) [inline csharp/src/SeedLang/X/SeedPython.cs:72]: Braces will be supported in SeedPython to initialize dictionary in the future. Shall we also add them in?

File: csharp/src/SeedLang.Shell/SourceCode.cs [hunk 3]
```diff
@@ -31,6 +33,17 @@ internal class SourceCode {
       _lines.Clear();
     }
 
+    internal void ParseAndWriteSource() {
+      // Tries to parse semantic tokens out of the source code. Falls back to syntax tokens if the
+      // source code is not valid.
+      if (!Executor.ParseSemanticTokens(Source, "", Language,
+                                        out IReadOnlyList<TokenInfo> syntaxTokens)) {
+        Executor.ParseSyntaxTokens(Source, "", Language,
+                                   out syntaxTokens);
```
Comment (codingpotato) [inline csharp/src/SeedLang.Shell/SourceCode.cs:42]: These two lines can be joined into one line.

File: csharp/src/SeedLang.Shell/Repl.cs [hunk 4]
```diff
@@ -29,6 +29,7 @@ internal sealed class Repl {
     internal Repl(SeedXLanguage language, RunType runType,
                   IEnumerable<VisualizerType> visualizerTypes) {
       _language = language;
```
Comment (codingpotato) [inline csharp/src/SeedLang.Shell/Repl.cs:31]: The variable `_language` can be removed and access the language by `_source.Language`.

**Review by codingpotato — CHANGES_REQUESTED** [review]

File: csharp/src/SeedLang/X/VisitorHelper.cs (hunk 1 above)
Comment (wixette) [reply csharp/src/SeedLang/X/VisitorHelper.cs]: Done

File: csharp/src/SeedLang/X/SeedPython.cs (hunk 2 above)
Comment (wixette) [reply csharp/src/SeedLang/X/SeedPython.cs]: Done. Added them.

**Review by codingpotato — APPROVED** [review]
LGTM

**Pull request merged**